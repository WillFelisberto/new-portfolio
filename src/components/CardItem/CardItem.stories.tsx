import { Meta, StoryObj } from '@storybook/react';
import CardItem from '.';

const meta: Meta<typeof CardItem> = {
  title: 'Components/CardItem',
  component: CardItem,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof CardItem>;

export const EducationExample: Story = {
  args: {
    logoUrl: '/ufsc.jpg',
    logoAlt: 'Education Logo',
    title: 'Computer Science',
    subtitle: 'Bachelor of Science',
    dateRange: 'Jan 2015 • Dec 2019 • (4 years)',
    description: {
      lexicalData: {
        root: {
          direction: 'ltr',
          format: '',
          indent: 0,
          type: 'root',
          version: 1,
          children: [
            {
              detail: 0,
              format: 0,
              mode: 'normal',
              style: '',
              text: 'Description',
              type: 'text',
              version: 1,
            },
          ],
        },
      },
    },
    linkUrl: 'https://university.edu',
    linkText: 'University Name',
  },
};

export const ExperienceExample: Story = {
  args: {
    logoUrl: '/ufsc.jpg',
    logoAlt: 'Company Logo',
    title: 'Company Name',
    subtitle: 'Software Engineer',
    dateRange: 'Feb 2020 • Present • (3 years)',
    description: {
      lexicalData: {
        root: {
          direction: 'ltr',
          format: '',
          indent: 0,
          type: 'root',
          version: 1,
          children: [
            {
              detail: 0,
              format: 0,
              mode: 'normal',
              style: '',
              text: 'Description',
              type: 'text',
              version: 1,
            },
          ],
        },
      },
    },
    linkUrl: 'https://company.com',
    linkText: '@ Company Name',
    additionalContent: (
      <div>
        <p className="mb-3 mt-6 text-sm font-semibold text-gray-400">Skills</p>
        <div className="mb-8 flex flex-wrap gap-x-2 gap-y-3 lg:max-w-[350px]">
          <span className="rounded bg-gray-600 px-2 py-1 text-white">
            React
          </span>
          <span className="rounded bg-gray-600 px-2 py-1 text-white">
            Node.js
          </span>
        </div>
      </div>
    ),
  },
};
