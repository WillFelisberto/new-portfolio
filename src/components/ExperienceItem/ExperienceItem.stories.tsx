import { Meta, StoryObj } from '@storybook/react';
import ExperienceItem, { WorkExperience } from '.';

const meta: Meta<typeof ExperienceItem> = {
  title: 'Components/ExperienceItem',
  component: ExperienceItem,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof ExperienceItem>;

const mockExperience: WorkExperience = {
  companyLogo: {
    url: '/ufsc.jpg',
  },
  role: 'Software Engineer',
  companyName: 'TechCorp',
  companyUrl: 'https://techcorp.com',
  startDate: '2020-01-01',
  endDate: '2023-01-01',
  technologies: [
    { name: 'React' },
    { name: 'Node.js' },
    { name: 'TypeScript' },
  ],
  description: {
    lexicalData: {
      root: {
        direction: 'ltr',
        format: '',
        indent: 0,
        type: 'root',
        version: 1,
        children: [
          {
            detail: 0,
            format: 0,
            mode: 'normal',
            style: '',
            text: 'Developed and maintained scalable web applications.',
            type: 'text',
            version: 1,
          },
        ],
      },
    },
  },
};

export const Default: Story = {
  args: {
    experience: mockExperience,
  },
};
