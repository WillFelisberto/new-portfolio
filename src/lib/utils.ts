import { ClassValue, clsx } from 'clsx';
import { twMerge } from 'tailwind-merge';
import { format, differenceInMonths, differenceInYears } from 'date-fns';
import { enUS } from 'date-fns/locale';

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

interface FormattedDateRange {
  formattedStartDate: string;
  formattedEndDate: string;
  formattedDuration: string;
}

export function formatDateRange(
  startDate: string,
  endDate: string | null,
): FormattedDateRange {
  const start = new Date(startDate);
  const end = endDate ? new Date(endDate) : new Date();

  const formattedStartDate = format(start, 'MMM yyyy', { locale: enUS });
  const formattedEndDate = endDate
    ? format(new Date(endDate), 'MMM yyyy', { locale: enUS })
    : 'Present';

  const months = differenceInMonths(end, start);
  const years = differenceInYears(end, start);
  const monthsRemaining = months % 12;

  const formattedDuration =
    years > 0
      ? `${years} year${years > 1 ? 's' : ''}${
          monthsRemaining > 0
            ? ` and ${monthsRemaining} month${monthsRemaining > 1 ? 's' : ''}`
            : ''
        }`
      : `${months} month${months > 1 ? 's' : ''}`;

  return { formattedStartDate, formattedEndDate, formattedDuration };
}
