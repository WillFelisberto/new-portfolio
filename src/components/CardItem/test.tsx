import { render } from '@testing-library/react';

import CardItem from '.';

describe('<CardItem />', () => {
  const mockProps = {
    logoUrl: '/path/to/logo.png',
    logoAlt: 'Example Logo',
    title: 'Example Title',
    subtitle: 'Example Subtitle',
    dateRange: 'Jan 2020 • Dec 2023 • (3 years)',
    description: 'description',
    linkUrl: 'https://example.com',
    linkText: 'Example Link',
    additionalContent: <div>Additional Content</div>,
  };
  it('should render card with all required props correctly', () => {
    const { getByText, getByLabelText, getByAltText, container } = render(
      <CardItem {...mockProps} />,
    );

    expect(getByAltText(/Example Logo/i)).toBeInTheDocument();
    expect(getByText(/Example Title/i)).toBeInTheDocument();
    expect(getByLabelText(/Example Link/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
