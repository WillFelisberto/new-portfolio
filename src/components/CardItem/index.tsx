/* eslint-disable @typescript-eslint/no-explicit-any */
'use client';
import { fadeUpAnimation } from '@/lib/animations';
import { motion } from 'framer-motion';
import Image from 'next/image';
import RichTextComponent from '../RichText';

export type CardItemProps = {
  logoUrl: string;
  logoAlt: string;
  title: string;
  dateRange: string;
  subtitle?: string;
  description?: any; // RichText content
  linkUrl?: string;
  education?: boolean;
  additionalContent?: React.ReactNode;
};

const CardItem = ({
  logoUrl,
  logoAlt,
  title,
  subtitle,
  dateRange,
  description,
  linkUrl,
  education = false,
  additionalContent,
}: CardItemProps) => {
  return (
    <motion.div
      className="grid grid-cols-[40px,1fr] gap-4 md:gap-10"
      {...fadeUpAnimation}
      transition={{ duration: 0.5 }}
    >
      <div className="flex flex-col items-center gap-4">
        {education ? (
          <div className="h-[60px] w-[60px] rounded-full border border-gray-500 bg-gray-50 p-0.5">
            <Image
              src={logoUrl}
              width={60}
              height={60}
              className="h-[60px] w-[60px] rounded-full"
              alt={`${title}`}
            />
          </div>
        ) : (
          <div className="rounded-full border border-gray-500 p-0.5">
            <Image
              src={logoUrl}
              width={40}
              height={40}
              className="rounded-full"
              alt={logoAlt}
            />
          </div>
        )}

        <div className="h-full w-[1px] bg-gray-800" />
      </div>

      <div>
        <div className="flex flex-col gap-2 text-sm sm:text-base">
          {linkUrl ? (
            <a
              href={linkUrl}
              target="_blank"
              aria-label={title}
              className="text-sky-100 transition-colors hover:text-sky-500"
              rel="noreferrer"
            >
              {title}
            </a>
          ) : (
            <h4 className="text-gray-300">{title}</h4>
          )}
          {subtitle && <p className="text-gray-400">{subtitle}</p>}
          <span className="text-gray-500">{dateRange}</span>
          {description && <RichTextComponent lexicalData={description} />}
        </div>

        {additionalContent && <div className="mt-4">{additionalContent}</div>}
      </div>
    </motion.div>
  );
};

export default CardItem;
