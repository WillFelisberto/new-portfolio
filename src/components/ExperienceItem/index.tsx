import { formatDateRange } from '@/lib/utils';
import CardItem from '../CardItem';
import TechBadge from '../TechBadge';
import {
  Media,
  Technology,
  Workexperience,
} from '@/app/(payload)/payload-types';

type ExperienceItemProps = {
  experience: Workexperience;
};

const ExperienceItem = ({ experience }: ExperienceItemProps) => {
  const {
    startDate,
    endDate,
    description,
    companyName,
    companyLogo,
    companyUrl,
    technologies,
  } = experience;

  const { formattedStartDate, formattedEndDate, formattedDuration } =
    formatDateRange(startDate!, endDate!);

  const additionalContent = (
    <>
      <p className="mb-3 mt-6 text-sm font-semibold text-gray-400">Skills</p>
      <div className="mb-8 flex flex-wrap gap-x-2 gap-y-3 lg:max-w-[350px]">
        {(technologies as Technology[]).map((tech, i) => (
          <TechBadge
            name={tech.name!}
            key={`experience-${companyName}-tech-${tech.name}`}
            transition={{ duration: 0.2, delay: i * 0.1 }}
          />
        ))}
      </div>
    </>
  );

  return (
    <CardItem
      logoUrl={(companyLogo as Media).url!}
      logoAlt={`Logo da empresa ${companyName}`}
      title={companyName!}
      subtitle={experience.role!}
      dateRange={`${formattedStartDate} • ${formattedEndDate} • (${formattedDuration})`}
      description={description}
      linkUrl={companyUrl!}
      additionalContent={additionalContent}
    />
  );
};

export default ExperienceItem;
