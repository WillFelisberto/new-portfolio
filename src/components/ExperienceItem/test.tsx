import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ExperienceItem, { WorkExperience } from '.';
import { CardItemProps } from '../CardItem';
import { formatDateRange } from '@/lib/utils';

jest.mock('../CardItem', () => ({
  __esModule: true,
  default: ({ title, subtitle, dateRange, linkText }: CardItemProps) => (
    <div>
      <h1>{title}</h1>
      <h2>{subtitle}</h2>
      <p>{dateRange}</p>
      <a href="#">{linkText}</a>
    </div>
  ),
}));

jest.mock('../TechBadge', () => ({
  __esModule: true,
  default: ({ name }: { name: string }) => <span>{name}</span>,
}));

jest.mock('@/lib/utils', () => ({
  formatDateRange: jest.fn(() => ({
    formattedStartDate: 'Jan 2020',
    formattedEndDate: 'Jan 2023',
    formattedDuration: '3 years',
  })),
}));

describe('<ExperienceItem />', () => {
  it('should render experience item with all required fields when valid experience data is provided', () => {
    const mockExperience: WorkExperience = {
      companyLogo: {
        url: '/ufsc.jpg',
      },
      role: 'Software Engineer',
      companyName: 'TechCorp',
      companyUrl: 'https://techcorp.com',
      startDate: '2020-01-01',
      endDate: '2023-01-01',
      technologies: [
        { name: 'React' },
        { name: 'Node.js' },
        { name: 'TypeScript' },
      ],
      description: {
        lexicalData: {
          root: {
            direction: 'ltr',
            format: '',
            indent: 0,
            type: 'root',
            version: 1,
            children: [
              {
                detail: 0,
                format: 0,
                mode: 'normal',
                style: '',
                text: 'Developed and maintained scalable web applications.',
                type: 'text',
                version: 1,
              },
            ],
          },
        },
      },
    };

    const { formattedStartDate, formattedEndDate, formattedDuration } =
      formatDateRange('2023-01-01', null);

    const { container } = render(
      <ExperienceItem experience={mockExperience} />,
    );

    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent(
      'TechCorp',
    );

    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent(
      'Software Engineer',
    );
    expect(screen.getByRole('paragraph')).toHaveTextContent(
      `${formattedStartDate} • ${formattedEndDate} • (${formattedDuration})`,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
