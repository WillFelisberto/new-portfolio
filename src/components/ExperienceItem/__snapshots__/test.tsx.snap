// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExperienceItem /> should render experience item with all required fields when valid experience data is provided 1`] = `
<div>
  <h1>
    TechCorp
  </h1>
  <h2>
    Software Engineer
  </h2>
  <p>
    Jan 2020 • Jan 2023 • (3 years)
  </p>
  <a
    href="#"
  />
</div>
`;
